### ISSUE TITLE
Some linting rules should be changed

### ISSUE THREAD

**Issue by trych** [basiljs/basil.js#159]
While linting the example scripts, I came across ~three~ ~four~ five linting rules that I think should be changed (i.e. turned off).

### [no-redeclare](http://eslint.org/docs/rules/no-redeclare)
This one turns out to be annoying as it does not allow you to have two simple for loops using `i` as their variable in your script. It makes it really cumbersome to use snippets for loops etc. Also it will be confusing to beginners (and pros?) to explain why a second for loop should look different than the first one (I am _not_ talking about nested loops here).

Example: [nameOnPage.jsx](https://github.com/basiljs/basil.js/blob/master/scripts/examples/document/nameOnPage.jsx).

![bildschirmfoto 2017-01-04 um 11 57 19](https://cloud.githubusercontent.com/assets/9803905/21640290/1f365b1a-d275-11e6-9982-9b675b917eef.png)

### [no-use-before-define](http://eslint.org/docs/rules/no-use-before-define)
This always gives an error when a function is defined at the end of the script and called somewhere above that. Quote from eslint.org: "[…] some believe it is best to always declare variables and functions before using them." I don't. 😉
Also, this seems to be an ES6 related rule, which should not concern us then.

Example: [guide.jsx](https://github.com/basiljs/basil.js/blob/master/scripts/examples/environment/guide.jsx)

![bildschirmfoto 2017-01-04 um 12 03 47](https://cloud.githubusercontent.com/assets/9803905/21640452/e5564184-d275-11e6-8488-f5b0f21b2630.png)

~### [guard-for-in](http://eslint.org/docs/rules/guard-for-in)~
~This enforces to filter results of for in loops with an if statement. Not sure why this should be enforced on the user, I don't think this should be turned on.~

~Example: [bounds.jsx](https://github.com/basiljs/basil.js/blob/master/scripts/examples/shape/bounds.jsx)~

![bildschirmfoto 2017-01-04 um 12 07 37](https://cloud.githubusercontent.com/assets/9803905/21640568/8d9a635c-d276-11e6-9bba-f3e988787875.png)

### [sort-vars](http://eslint.org/docs/rules/guard-for-in)
Quote eslint.org: "some developers prefer to sort variable names alphabetically". Again, I don't, I think sometimes sorting `myMin`, `myMax` makes more sense. Let's turn this off.

Example: [04_01.jsx](https://github.com/basiljs/basil.js/blob/master/scripts/tutorials/basic/04_Remixer/04_1.jsx) (this one needs some fixing of the line breaks btw.)

![bildschirmfoto 2017-01-04 um 12 23 02](https://cloud.githubusercontent.com/assets/9803905/21640940/ca64f5ac-d278-11e6-8618-8b5cf5eca6fa.png)

### [radix](http://eslint.org/docs/rules/radix)
This just seems unnecessary to me.

Example: [08_01.jsx](https://github.com/basiljs/basil.js/blob/master/scripts/tutorials/advanced/08_bar_chart_from_csv_data/08_1.jsx)

![bildschirmfoto 2017-01-04 um 12 31 06](https://cloud.githubusercontent.com/assets/9803905/21641145/420633a4-d27a-11e6-9a22-a19586220aa1.png)

Can we turn those off? Pleeeeease? :)

**Comment by fabianmoronzirfas** [comment]
I'm actually like all these rules.  😄 

Why you should guard loops --> http://stackoverflow.com/questions/1963102/what-does-the-jslint-error-body-of-a-for-in-should-be-wrapped-in-an-if-statemen
 youWhy should not use `for in` loops on Arrays. http://stackoverflow.com/questions/500504/why-is-using-for-in-with-array-iteration-a-bad-idea

**Comment by trych** [comment]
Ok, actually the third one makes sense to replace with an simple for loop altogether, I guess.

@fabianmoronzirfas Can you explain why you like the first two rules? Should we write functions like this:

```js
// @includepath "~/Documents/;%USERPROFILE%Documents";
// @include "basiljs/basil.js";

// user functions, maybe a lot of them
function doThing1() {}
function doThing2() {}
function doThing3() {}
function doThing4() {}

// at the bottom of the file the actual main draw function
function draw() {

doThing1();
doThing2();
doThing3();
doThing4();

}

b.go();
```

and for loops like this:

```js
// @includepath "~/Documents/;%USERPROFILE%Documents";
// @include "basiljs/basil.js";

function draw() {

for (var i = 0; i < Things.length; i++) {
  // do stuff
}

// do more stuff

for (var j = 0; j < Things.length; j++) {
  // do stuff
}

// do more stuff

for (var k = 0; k < Things.length; k++) {
  // do stuff
}

}

b.go();
```

? I don't see how this is helpful, especially not in example scripts for beginners. This only adds confusion.

**Comment by fabianmoronzirfas** [comment]
On the first: I just like my code sorted. If I read it from top I first see the functions and at the end I have an execution of all that. 

On the second one: Your example makes the overview simple. If I have lots of code within the loops and maybe some switch and if statements it is hard to track in which loop I am when trying to get a quick overview. I will have to count the closing `}` if all use `i` as iterator

**Comment by trych** [comment]
Ok, I put it differently: I can understand why you _like_ the rules (I mean I know they were invented for a reason), but I still think we should check if they make sense in the context. The context here is example scripts for absolute beginners.

As I have been an absolute beginner not too long ago, I can assure you that it would totally confuse me, if an example script did not always use `i` as the iterator. As a beginner I just think:

Ah, now I need to loop something, so I use the standard for loop I learned about, using i. And then later: Ah, I need to loop something else, so I will again use the for loop I learned about using i.

About the function order: Hm, I think this comes down to preference. I like my helper functions at the end and the main function at the top. I would like to hear the opinion of the others, so we can see if there is a preference in the team. @b-g @ffd8 ?

@fabianmoronzirfas How about an extra .eslintrc file for the example scripts and tutorials? I think the special circumstances (scripts for absolute beginners) require linting that maybe does not accord to best practice JavaScript, but makes sense for those beginners.

**Comment by fabianmoronzirfas** [comment]
> Ah, now I need to loop something, so I use the standard for loop I learned about, using i. And then later: Ah, I need to loop something else, so I will again use the for loop I learned about using i.  

I think for beginners it is important to learn that using `i`is just a convention not mandatory.  

> How about an extra .eslintrc file for the example scripts and tutorials?  

I would say ~~noway~~ okay. The `draw` and `setup` at the top is okay for me. The redeclare should stay IMO.

**Comment by trych** [comment]
> The redeclare should stay IMO.

Ok, so what is the convention you would use then? `j` for the second loop, `k` for the third loop and so on?

Also, what do you think about rules 4 or 5 in my edited opening post? (Rule 5 really does not matter that much I guess, 4 I would quite like to turn off).

**Comment by fabianmoronzirfas** [comment]
Sorting vars: Yes this is for quite pedantic people. 🤓  Should not matter in the examples. In the core I would like to see this as well.

Radix operator: Can be omitted for the examples. Should be there for the core.

**Comment by trych** [comment]
Ok, cool, sounds good to me.
So I'll prepare a PR that introduces a `.eslintrc` file in `scripts/` that turns off rule 2, 4 and 5.

One more thing: `HashList` was not added to globals yet (in the existing `.eslintrc`), so it threw a linting error, too. Will add it to the globals list.

**Comment by b-g** [comment]
Closed as already decided with last PR #160 :)

My feeling with all this is a bit ambivalent: a tidy code base in the core is important and super ... and clean examples are nice, but non-linted would have been fine for me as well, if we had put the energy in new examples. Long story short, please don't go too crazy with the linting in the non-core code realm :)
